internal/dockerauth/api.go

```diff
@@ -107,23 +111,133 @@ type tokenResponse struct {
 
 type dockerRegistryClaims struct {
 	jwt.StandardClaims
-	Access []ResourceAccessRights `json:"access"`
+	Access []resourceAccessRights `json:"access"`
 }
 
-func (h *Handler) Token(p httprequest.Params, req *tokenRequest) (*tokenResponse, error) {
-	ras, err := ParseScope(req.Scope)
+func (h *Handler) handler(p httprequest.Params) (*handler, context.Context, error) {
+	store, err := h.params.Pool.RequestStore()
+	if err != nil {
+		return nil, p.Context, errgo.Mask(err)
+	}
+	return &handler{h, store}, p.Context, nil
+}
+
+type handler struct {
+	h     *Handler
+	store *charmstore.Store
+}
+
+func (h *handler) Close() error {
+	h.store.Close()
+	return nil
+}
+
+func (h *handler) Token(p httprequest.Params, req *tokenRequest) (*tokenResponse, error) {
+	ras, err := parseScope(req.Scope)
 	if err != nil {
 		logger.Infof("Invalid scope: %s", err)
 	}
-	var issuer string
-	if len(h.Certs) > 0 {
-		issuer = h.Certs[0].Subject.CommonName
+	ms := credentials(p.Request)
+	declared := checkers.InferDeclared(ms)
+	var user user = noUser{}
+	if identity, err := h.h.params.IDMClient.DeclaredIdentity(declared); err == nil {
+		if u, ok := identity.(*idmclient.User); ok {
+			user = u
+		}
+	}
+	filteredRAs := make([]resourceAccessRights, 0, len(ras))
+	for _, ra := range ras {
+		if ra.Type != "repository" {
+			continue
+		}
+		acl := h.repositoryACL(ra.Name)
+		filteredActions := make([]string, 0, len(ra.Actions))
+		for _, a := range ra.Actions {
+			switch a {
+			case "pull":
+				allow, _ := user.Allow(acl.Read)
```

Comment:
I'm not convinced that we should be ignoring all these errors. At the very least we should be logging them, but arguably if we can't get ACLs, we should return the error. A similar thing applies to parseScope above. These aren't errors that should happen in the normal course of things.